common/pldm/src/message/firmware_update/activate_fw.rs

```diff
@@ -0,0 +1,137 @@
+// Licensed under the Apache-2.0 license
+
+use crate::codec::{PldmCodec, PldmCodecError};
+use crate::protocol::base::{
+    InstanceId, PldmMsgHeader, PldmMsgType, PldmSupportedType, PLDM_MSG_HEADER_LEN,
+};
+use crate::protocol::firmware_update::FwUpdateCmd;
+use zerocopy::{FromBytes, Immutable, IntoBytes};
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+#[repr(u8)]
+pub enum SelfContainedActivationRequest {
+    NotActivateSelfContainedComponents = 0,
+    ActivateSelfContainedComponents = 1,
+}
+
+#[derive(Debug, Clone, FromBytes, IntoBytes, Immutable, PartialEq)]
+#[repr(C, packed)]
+pub struct ActivateFirmwareRequest {
+    pub hdr: PldmMsgHeader<[u8; PLDM_MSG_HEADER_LEN]>,
+    pub self_contained_activation_req: u8,
+}
+
+impl ActivateFirmwareRequest {
+    pub fn new(
+        instance_id: InstanceId,
+        msg_type: PldmMsgType,
+        self_contained_activation_req: SelfContainedActivationRequest,
+    ) -> ActivateFirmwareRequest {
+        ActivateFirmwareRequest {
+            hdr: PldmMsgHeader::new(
+                instance_id,
+                msg_type,
+                PldmSupportedType::FwUpdate,
+                FwUpdateCmd::ActivateFirmware as u8,
+            ),
+            self_contained_activation_req: self_contained_activation_req as u8,
+        }
+    }
+}
+
+impl PldmCodec for ActivateFirmwareRequest {
+    fn encode(&self, buffer: &mut [u8]) -> Result<usize, PldmCodecError> {
+        let bytes = core::mem::size_of::<ActivateFirmwareRequest>();
+        if buffer.len() < bytes {
+            return Err(PldmCodecError::BufferTooShort);
+        };
+
+        self.write_to(&mut buffer[..bytes]).unwrap();
+        Ok(bytes)
+    }
+
+    fn decode(buffer: &[u8]) -> Result<Self, PldmCodecError> {
+        let bytes = core::mem::size_of::<ActivateFirmwareRequest>();
+        if buffer.len() < bytes {
```

Comment:
Similarly I think this could be simplified a little with `read_from_prefix` and `map_err` / `map`